blocks/monitoring/__init__.py

```diff
@@ -0,0 +1,193 @@
+"""This part of the framework helps you monitor your model during training.
+
+:class:`MonitorChannels` describe quantities that should be monitored. They
+can either be Theano variables (e.g. the objective function, the weight
+norms, etc.) or they can be callable functions (e.g. if you want to use
+NumPy to calculate eigenvalues, or sample text from your language model).
+
+.. todo::
+
+   Monitored Theano variables might depend on inputs beyond those given to
+   the model. There should be a way of providing these to the monitoring
+   channels, collecting them, and compiling them as part of the Theano
+   function that performs monitoring.
+
+"""
+import logging
+from abc import ABCMeta, abstractmethod
+
+from blocks.utils import shared_expression, update_instance
+
+logger = logging.getLogger(__name__)
+
+
+class MonitorChannel(object):
```

Comment:
I'm not sure what the best alternative would be. `Aggregator` doesn't make any sense to me, because `ModelProperty` doesn't aggregate anything, and in many cases the variables monitored will be of this type. I think it will be very counter-intuitive for a user to be asked to use an `Aggregator` in order to monitor static model properties such as weight norms.

Something like *statistics* or *info* as suggested by @rizar could make sense. But taking the weight norm, `W_norm = tensor.mean(W)`, and then saying that we provide a "statistic" of this variable is slightly counter intuitive for me as well (it would suggest you are providing the mean, variance, etc. of this variable).

I find monitor channel a better option already, because it's simply a channel that *can* be monitored (which is what the single purpose of this entire class is after all, to provide the infrastructure that allows us to monitor things). The fact that it's called a monitor channel (as opposed to monitored channel, or monitored variable) doesn't necessarily imply that it *is* monitored, merely that it can be.

The only thing the monitoring extension now has to do is collect the aggregation schemes and apply them to different datasets to get a set of monitoring channels. These monitoring channels can then be monitored and their results added to the log. I don't think this would confuse anyone.